### PR TITLE
Custom Logging Format

### PR DESCRIPTION
**Pull request by BattleCh1cken**
#### Summary

This pull request adds support for custom logging formats to LemLib.

#### Motivation

This pull request is being created to make it possible for users to make graphical debuggers for odometry and PID. The new format API allows the user to customize LemLib's logging output to whatever their needs are.

#### Test Plan

The code needs to be run on a bot and each of the new logging methods need to be tested, along with changing the format.

#### Additional Notes

This PR is completely untested, and is therefore likely to change. I'd also like to give the user control over the refresh rate of the logger, but I need to do more testing to figure out how that will work.

This pull request is also completely undocumented. I will get that done as soon as I finish testing.

### CODE REVIEW

**Comment by BattleCh1cken** [comment]
This PR has been now tested and works. The only thing that I'm unsure about it the way that other parts of the program access the logger. Currently I have a globally defined object defined the the logging header, but I don't know if this is a good idea.

A viable alternative could be making the user define the logger, and then taking a pointer to it in the chassiss's constructor. Note that this would also require taking it as input in the FAPID constructor as well.

File: src/lemlib/logger.cpp
Comment (liam-teale) [inline src/lemlib/logger.cpp:1]: This file needs some more comments in the source. E.x why are you using a deque? What does the buffer do?

File: src/lemlib/pid.cpp [hunk 1]
```diff
@@ -93,6 +93,8 @@ float lemlib::FAPID::update(float target, float position, bool log) {
     prevOutput = output;
     prevError = error;
     totalError += error;
+
+    if (log) { Logger::logPid(name, output, kI * error, kD * totalError, kD * deltaError); }
```
Comment (liam-teale) [inline src/lemlib/pid.cpp]: This line computes the outputs incorrectly
```suggestion
    if (log) { Logger::logPid(name, output, kP * error, kI * totalError, kD * deltaError); }
```

File: include/lemlib/logger.hpp [hunk 2]
```diff
@@ -11,129 +11,180 @@
...
+        /*
+         * @brief Logs the output of a PID controller.
+         *
+         * @param name The name of the controller
+         * @param output The total output of the controller which consists of all of the other terms added together.
+         * @param p The proportional term of the controller
+         * @param i The integral term of the controller
+         * @param d The derivative term of the controller
+         */
+        static void logPid(std::string name, float output, float p, float i, float d);
```
Comment (liam-teale) [inline include/lemlib/logger.hpp]: Error should be logged as well

File: include/lemlib/logger.hpp [hunk 3]
```diff
@@ -11,129 +11,180 @@
...
+         * @param rate The new print rate in milliseconds.
+         */
+        static void setPrintRate(int rate);
 
-static Level lowestLevel = Level::INFO;
+        /*
+         * @brief Begins the long running task for the logger.
+         * @note This method must be run in order for the logger to function.
+         */
+        static void startTask();
```
Comment (liam-teale) [inline include/lemlib/logger.hpp]: should be renamed
```suggestion
        static void initialize();
```

File: src/main.cpp [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #include "main.h"
 #include "lemlib/api.hpp"
+#include "lemlib/logger.hpp"
```
Comment (liam-teale) [inline src/main.cpp]: Unnecessary include, logger.hpp is already included in api.hpp

**Review by liam-teale — CHANGES_REQUESTED** [review]

**Review by liam-teale — APPROVED** [review]

**Pull request merged**